server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.server.jetty.jsplugin;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.jimfs.Configuration;
+import com.google.common.jimfs.Jimfs;
+import io.deephaven.plugin.type.JsPlugin;
+import io.deephaven.plugin.type.JsPluginRegistration;
+import org.eclipse.jetty.servlet.DefaultServlet;
+import org.eclipse.jetty.servlet.ServletHolder;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link JsPluginRegistration} for Jetty.
+ *
+ * @see #servletHolder(String)
+ */
+public final class JsPlugins implements JsPluginRegistration {
+
+    /**
+     * Creates a new js plugins instance with an in-memory filesystem.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final FileSystem fs = Jimfs.newFileSystem(Configuration.unix());
+        final JsPlugins jsPlugins = new JsPlugins(fs.getPath("/"));
+        jsPlugins.writeManifest(); // empty manifest
+        return jsPlugins;
+    }
+
+    private final Path path;
+    private final Map<String, JsPlugin> plugins;
+
+    private JsPlugins(Path path) {
+        this.path = Objects.requireNonNull(path);
+        this.plugins = new LinkedHashMap<>();
+    }
+
+    /**
+     * Creates a {@link DefaultServlet} servlet holder.
+     *
+     * @param name the name
+     * @return the servlet holder
+     */
+    public ServletHolder servletHolder(String name) {
+        final ServletHolder jsPlugins = new ServletHolder(name, DefaultServlet.class);
+        jsPlugins.setInitParameter("resourceBase", path.toUri().toString());
+        jsPlugins.setInitParameter("pathInfoOnly", "true");
+        jsPlugins.setInitParameter("dirAllowed", "false");
+        jsPlugins.setAsyncSupported(true);
+        return jsPlugins;
+    }
+
+    @Override
+    public synchronized void register(JsPlugin jsPlugin) {
+        if (plugins.containsKey(jsPlugin.name())) {
+            throw new IllegalArgumentException(
+                    String.format("js plugin with name '%s' already exists", jsPlugin.name()));
+        }
+        try {
+            jsPlugin.copyTo(path.resolve(jsPlugin.name()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        plugins.put(jsPlugin.name(), jsPlugin);
```

Comment:
As written, `writeManifest()` depends on `plugins`. We _could_ modify it to `writeManifest(jsPlugin)`, but that seems a bit ill-structured just to suit.

I'm simplifying the writeManifest a bit, but *think* the order makes sense as-is.

I don't expect writeManifest to fail.

Reply:
Should somehow match that perspective in code then? "This can't possibly fail, so we're going to make sure the system is unusable if it somehow does manage to fail". Since you'd like the rest of the impl to be agnostic of exactly which FileSystem impl is in use, it should be expected that an ordinary write could fail for some reason. If we're actually assuming that jimfs is the only option, I'd agree with you here, but ask you make  the API hard/impossible to use (even in unit tests) with a different fs.

Reply:
In the current path to registrations, if the registration fails, startup will fail (explicitly made the permissions on a file bad):

```
2022-10-18T17:50:05.354Z | main                 |  INFO | .d.s.p.PluginRegistration | Registering js plugin: @deephaven/js-plugin-plotly@0.1.0 / io.deephaven.server.plugin.type.JsPluginDistribution@2c413ffc                   
Initiating shutdown due to: Uncaught exception in thread main                                                                                                                                                                    
java.lang.RuntimeException: Error in Python interpreter:                                                                                                                                                                         
Type: <class 'RuntimeError'>                                                                                                                                                                                                     
Value: java.nio.file.AccessDeniedException: /tmp/dh/lib64/python3.10/site-packages/deephaven_plugin_plotly_js/dist/index.js                                                                                                      
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)                                                                                                                                      
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)                                                                                                                                       
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)                                                                                                                                       
        at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)                                                                                                                           
        at java.base/java.nio.file.Files.newByteChannel(Files.java:371)                                                                                                                                                          
        at java.base/java.nio.file.Files.newByteChannel(Files.java:422)                                                                                                                                                          
        at java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:420)                                                                                                                            
        at java.base/java.nio.file.Files.newInputStream(Files.java:156)                                                                                                                                                          
        at java.base/java.nio.file.CopyMoveHelper.copyToForeignTarget(CopyMoveHelper.java:125)                                                                                                                                   
        at java.base/java.nio.file.Files.copy(Files.java:1298)                                                                                                                                                                   
        at io.deephaven.server.plugin.type.JsPluginDistribution$CopyRecursiveVisitor.visitFile(JsPluginDistribution.java:141)                                                                                                    
        at io.deephaven.server.plugin.type.JsPluginDistribution$CopyRecursiveVisitor.visitFile(JsPluginDistribution.java:122)                                                                                                    
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2725)                                                                                                                                                           
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2797)                                                                                                                                                           
        at io.deephaven.server.plugin.type.JsPluginDistribution.copyRecursive(JsPluginDistribution.java:103)                                                                                                                     
        at io.deephaven.server.plugin.type.JsPluginDistribution.copyTo(JsPluginDistribution.java:98)                                                                                                                             
        at io.deephaven.server.jetty.jsplugin.JsPlugins.register(JsPlugins.java:93)                                                                                                                                              
        at io.deephaven.server.plugin.PluginRegistrationVisitor.visit(PluginRegistrationVisitor.java:44)                                                                                                                         
        at io.deephaven.server.plugin.PluginRegistrationVisitor.visit(PluginRegistrationVisitor.java:17)                                                                                                                         
        at io.deephaven.plugin.type.JsPluginBase.walk(JsPluginBase.java:13)                                                                                                                                                      
        at io.deephaven.server.plugin.PluginRegistrationVisitor.register(PluginRegistrationVisitor.java:32)                                                                                                                      
        at io.deephaven.server.plugin.PluginRegistration$Counting.visit(PluginRegistration.java:73)                                                                                                                              
        at io.deephaven.server.plugin.PluginRegistration$Counting.visit(PluginRegistration.java:46)                                                                                                                              
        at io.deephaven.plugin.type.JsPluginBase.walk(JsPluginBase.java:13)                                                                                                                                                      
        at io.deephaven.server.plugin.PluginRegistration$Counting.register(PluginRegistration.java:53)                                                                                                                           
        at io.deephaven.server.plugin.python.CallbackAdapter.registerJsPlugin(CallbackAdapter.java:31)                                                                                                                           
        at org.jpy.PyLib.callAndReturnValue(Native Method)                                                                                                                                                                       
        at org.jpy.PyProxyHandler.invoke(PyProxyHandler.java:120)                                                                                                                                                                
        at io.deephaven.server.plugin.python.$Proxy5.initialize_all_and_register_into(Unknown Source)                                                                                                                            
        at io.deephaven.server.plugin.python.PythonPluginsRegistration.registerInto(PythonPluginsRegistration.java:33)                                                                                                           
        at io.deephaven.server.plugin.PluginRegistration.registerAll(PluginRegistration.java:41)                                                                                                                                 
        at io.deephaven.server.runner.DeephavenApiServer.run(DeephavenApiServer.java:136)
        at io.deephaven.server.jetty.JettyMain.main(JettyMain.java:25)
```